[SystemZ][z/OS] Add new openFileForReadBinary function, and pass IsText parameter to getBufferForFile

abhina-sree opened the pull request:
This patch adds an IsText parameter to the following getBufferForFile, getBufferForFileImpl. We introduce a new virtual function openFileForReadBinary which defaults to openFileForRead except in RealFileSystem which uses the OF_None flag instead of OF_Text.

The default is set to OF_Text instead of OF_None, this change in value does not affect any other platforms other than z/OS. Setting this parameter correctly is required to open files on z/OS in the correct encoding. The IsText parameter is based on the context of where we open files, for example, in the ASTReader, HeaderMap requires that files always be opened in binary even though they might be tagged as text.

abhina-sree committed:
[SystemZ][z/OS] Add new openFileForReadBinary function, and pass IsText parameter to getBufferForFile

perry-ca commented on llvm/lib/Support/VirtualFileSystem.cpp:
```diff
@@ -117,8 +117,12 @@ FileSystem::~FileSystem() = default;
-  auto F = openFileForRead(Name);
+                             bool RequiresNullTerminator, bool IsVolatile,
+                             bool IsText) {
+  auto openFileFunctionPointer = &FileSystem::openFileForRead;
+  if (!IsText)
+    openFileFunctionPointer = &FileSystem::openFileForReadBinary;
```
```suggestion
  auto F = IsText ?openFileForRead(Name) : openFileForReadBinary(Name);
```

abhina-sree replied:
Thanks this is done

perry-ca commented on llvm/lib/Support/VirtualFileSystem.cpp:
```diff
@@ -324,6 +330,17 @@ ErrorOr<Status> RealFileSystem::status(const Twine &Path) {
 
 ErrorOr<std::unique_ptr<File>>
 RealFileSystem::openFileForRead(const Twine &Name) {
+  SmallString<256> RealName, Storage;
```
Can you put this code into a static function or even a private member function so you can avoid the duplication of code.

abhina-sree replied:
I've created a private function for both functions to call, thanks!

perry-ca commented on llvm/lib/Support/VirtualFileSystem.cpp:
```diff
@@ -302,6 +305,17 @@ class RealFileSystem : public FileSystem {
     return Storage;
   }
 
+  ErrorOr<std::unique_ptr<File>> openFileForRead(const Twine &Name,
```
Thanks, Could you mark this as static.

abhina-sree replied:
I don't think I'm able to mark it static because it calls adjustPath which is non-static

perry-ca replied:
np.  I was misreading the diff.  I didn't see it was a private member function.  It looked like a file scope function.  I was thinking file scope static, not static member.

kadircet commented on llvm/lib/Support/VirtualFileSystem.cpp:
```diff
@@ -302,6 +305,17 @@ class RealFileSystem : public FileSystem {
     return Storage;
   }
 
+  ErrorOr<std::unique_ptr<File>> openFileForRead(const Twine &Name,
```
it was quite confusing to see this as an overload, can you call it `openFileForReadWithFlags` ?

abhina-sree replied:
Thanks, this is done

kadircet commented on llvm/include/llvm/Support/VirtualFileSystem.h:
```diff
@@ -271,15 +271,24 @@ class FileSystem : public llvm::ThreadSafeRefCountedBase<FileSystem>,
   virtual llvm::ErrorOr<std::unique_ptr<File>>
   openFileForRead(const Twine &Path) = 0;
 
+  /// Get a \p File objct for the binary file at \p Path, if one exists.
+  /// This function should be called instead of openFileForRead if the file
+  /// should be opened as a binary file.
```
can you rather explain the motiviating use case? e.g. "Some filesystems can perform modifications to bytes on-disk when reading as test. Reading as binary ensures contents are received as-is in such cases, on most file systems this is same as `openFileForRead`".

abhina-sree replied:
Thanks, I've updated the comments, please let me know if they need further improvement.

kadircet commented on llvm/include/llvm/Support/VirtualFileSystem.h:
```diff
@@ -271,15 +271,24 @@ class FileSystem : public llvm::ThreadSafeRefCountedBase<FileSystem>,
   getBufferForFile(const Twine &Name, int64_t FileSize = -1,
-                   bool RequiresNullTerminator = true, bool IsVolatile = false);
+                   bool RequiresNullTerminator = true, bool IsVolatile = false,
+                   bool IsText = true);
```
can you also add some comments about what `IsText` parameter does?

abhina-sree replied:
I've added a sentence, please let me know if it needs more explanation, thanks!

kadircet commented on clang/lib/Lex/HeaderMap.cpp:
```diff
@@ -54,7 +54,9 @@ std::unique_ptr<HeaderMap> HeaderMap::Create(FileEntryRef FE, FileManager &FM) {
-  auto FileBuffer = FM.getBufferForFile(FE);
+  auto FileBuffer =
+      FM.getBufferForFile(FE, /*IsVolatile=*/false,
+                          /*RequiresNullTerminator=*/true, /*IsText=*/false);
```
i think it'd be great to have a test case for this that actually fails (even if only in zOS).

abhina-sree replied:
So the following tests were previously failing on z/OS due to the binary/text issue and will be fixed by this change
clang/test/Preprocessor/headermap-rel.c
clang/test/Preprocessor/headermap-rel2.c

kadircet commented on clang/include/clang/Basic/FileManager.h:
```diff
@@ -292,21 +292,21 @@ class FileManager : public RefCountedBase<FileManager> {
   llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>
   getBufferForFile(FileEntryRef Entry, bool isVolatile = false,
-                   bool RequiresNullTerminator = true,
+                   bool RequiresNullTerminator = true, bool IsText = true,
```
can you add some comments explaining what `IsText` is about and when it should be set to `false`?

also can you make it the last parameter ?

abhina-sree replied:
I've added a brief comment here as well, please let me know if its sufficient or needs more, thanks!

kadircet reviewed (changes requested):
thanks! I think this LG at a high-level, added some comments about the details.

I think it'd be worthwhile to add some tests (even if not now, probably going forward). as otherwise such rare use cases might regress quite frequently.

cor3ntin commented:
@abhina-sree do you plan to modify embed in this PR?

abhina-sree committed:
address comments

abhina-sree commented:
> @abhina-sree do you plan to modify embed in this PR?

The related lit tests all seem to be passing on z/OS right now. I'm guessing it might be because we get the contents as binary here https://github.com/llvm/llvm-project/blob/main/clang/lib/Lex/PPDirectives.cpp#L3990. I'll have to do a bit more investigation, but it seems to be working without modifications so I don't think any changes are needed. If there are changes required, I will make them in a separate PR

abhina-sree commented:
> @abhina-sree do you plan to modify embed in this PR?

Actually Sean has pointed out that if we pass in non-ascii inputs, the tests do in fact fail. So I will look into passing the IsText parameter in the necessary functions. Thanks for pointing this out!

kadircet reviewed (approved):
thanks for bearing with me!

abhina-sree committed:
Pass IsText from getFileRef to open files as binary for #embed and add testcase for zos

abhina-sree commented:
@cor3ntin, I've added support for #embed handling now, I've added a z/OS only testcase clang/test/Preprocessor/embed_zos.c which is a copy of clang/test/Preprocessor/embed_art.c except it changes the file tag to be non-ascii which is now working with the latest commit.
